File: python/test_api/api_impl/test_api0.py
from ..api_api import api
from ..api_api.shared import EventHook
#from api_api import api
#from api_api.shared import EventHook
from typing import Iterable

class TestApi0(api.ITestApi0):
    def __init__(self):
        super().__init__()
        self._prop_int: int = 0
        self._prop_float: float = 0.0
        self._prop_string: str = ""
        self.on_prop_int_changed: int = EventHook()
        self.on_prop_float_changed: float = EventHook()
        self.on_prop_string_changed: str = EventHook()
        self.on_sig_int = EventHook()
        self.on_sig_float = EventHook()
        self.on_sig_string = EventHook()

    def set_prop_int(self, value):
        if self._prop_int == value:
            return
        self._prop_int = value
        self._push_prop_int(self._prop_int)
    
    def get_prop_int(self):
        return self._prop_int        

    def _push_prop_int(self, value):
        self.on_prop_int_changed.fire(value)

    def set_prop_float(self, value):
        if self._prop_float == value:
            return
        self._prop_float = value
        self._push_prop_float(self._prop_float)
    
    def get_prop_float(self):
        return self._prop_float        

    def _push_prop_float(self, value):
        self.on_prop_float_changed.fire(value)

    def set_prop_string(self, value):
        if self._prop_string == value:
            return
        self._prop_string = value
        self._push_prop_string(self._prop_string)
    
    def get_prop_string(self):
        return self._prop_string        

    def _push_prop_string(self, value):
        self.on_prop_string_changed.fire(value)

    def func_int(self, param_int: int) -> int:
        return param_int

    def func_float(self, param_float: float) -> float:
        return 0.0

    def func_string(self, param_string: str) -> str:
        return ""

    def _sig_int(self, param_int: int):
        self.on_sig_int.fire(param_int)

    def _sig_float(self, param_float: float):
        self.on_sig_float.fire(param_float)

    def _sig_string(self, param_string: str):
        self.on_sig_string.fire(param_string)

File: cpp/testApi/modules/api/implementation/testapi0.cpp


#include "api/implementation/testapi0.h"
#include "api/generated/core/testapi0.publisher.h"
#include "api/generated/core/testapi0.data.h"

using namespace Cpp::Api;

TestApi0::TestApi0()
    : m_publisher(std::make_unique<TestApi0Publisher>())
{
}
TestApi0::~TestApi0()
{
}

void TestApi0::setPropInt(int propInt)
{
    if (m_data.m_propInt != propInt) {
        m_data.m_propInt = propInt;
        m_publisher->publishPropIntChanged(propInt);
    }
}

int TestApi0::getPropInt() const
{
    return m_data.m_propInt;
}

void TestApi0::setPropFloat(float propFloat)
{
    if (m_data.m_propFloat != propFloat) {
        m_data.m_propFloat = propFloat;
        m_publisher->publishPropFloatChanged(propFloat);
    }
}

float TestApi0::getPropFloat() const
{
    return m_data.m_propFloat;
}

void TestApi0::setPropString(const std::string& propString)
{
    if (m_data.m_propString != propString) {
        m_data.m_propString = propString;
        m_publisher->publishPropStringChanged(propString);
    }
}

const std::string& TestApi0::getPropString() const
{
    return m_data.m_propString;
}

int TestApi0::funcInt(int paramInt)
{
    return paramInt; 
}

std::future<int> TestApi0::funcIntAsync(int paramInt, std::function<void(int)> callback)
{
    return std::async(std::launch::async, [this, callback,
                    paramInt]()
        {auto result = funcInt(paramInt);
            if (callback)
            {
                callback(result);
            }return result;
        }
    );
}

float TestApi0::funcFloat(float paramFloat)
{
    (void) paramFloat; // suppress the 'Unreferenced Formal Parameter' warning.
    // do business logic here
    return 0.0f;
}

std::future<float> TestApi0::funcFloatAsync(float paramFloat, std::function<void(float)> callback)
{
    return std::async(std::launch::async, [this, callback,
                    paramFloat]()
        {auto result = funcFloat(paramFloat);
            if (callback)
            {
                callback(result);
            }return result;
        }
    );
}

std::string TestApi0::funcString(const std::string& paramString)
{
    (void) paramString; // suppress the 'Unreferenced Formal Parameter' warning.
    // do business logic here
    return std::string();
}

std::future<std::string> TestApi0::funcStringAsync(const std::string& paramString, std::function<void(std::string)> callback)
{
    return std::async(std::launch::async, [this, callback,
                    paramString]()
        {auto result = funcString(paramString);
            if (callback)
            {
                callback(result);
            }return result;
        }
    );
}

void TestApi0::funcVoid(int paramInt)
{
    (void) paramInt; // suppress the 'Unreferenced Formal Parameter' warning.
    // do business logic here
}

std::future<void> TestApi0::funcVoidAsync(int paramInt, std::function<void(void)> callback)
{
    return std::async(std::launch::async, [this, callback,
                    paramInt]()
        {funcVoid(paramInt);
            if (callback)
            {
                callback();
            }
        }
    );
}

ITestApi0Publisher& TestApi0::_getPublisher() const
{
    return *m_publisher;
}
